server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/IntValuesComparatorSource.java

```diff
@@ -0,0 +1,144 @@
+
+    public IntValuesComparatorSource(
+        IndexNumericFieldData indexFieldData,
+        @Nullable Object missingValue,
+        MultiValueMode sortMode,
+        Nested nested
+    ) {
+        super(missingValue, sortMode, nested);
+        this.indexFieldData = indexFieldData;
+    }
+
+    @Override
+    public SortField.Type reducedType() {
+        return SortField.Type.INT;
+    }
+
+    private NumericDocValues getNumericDocValues(LeafReaderContext context, int missingValue) throws IOException {
+        final SortedNumericDocValues values = indexFieldData.load(context).getLongValues();
+        if (nested == null) {
+            return FieldData.replaceMissing(sortMode.select(values), missingValue);
+        }
+        final BitSet rootDocs = nested.rootDocs(context);
+        final DocIdSetIterator innerDocs = nested.innerDocs(context);
+        final int maxChildren = nested.getNestedSort() != null ? nested.getNestedSort().getMaxChildren() : Integer.MAX_VALUE;
+        return sortMode.select(values, missingValue, rootDocs, innerDocs, context.reader().maxDoc(), maxChildren);
+    }
+
+    @Override
+    public FieldComparator<?> newComparator(String fieldname, int numHits, boolean enableSkipping, boolean reversed) {
+        assert indexFieldData == null || fieldname.equals(indexFieldData.getFieldName());
+
+        final int iMissingValue = (Integer) missingObject(missingValue, reversed);
+        // NOTE: it's important to pass null as a missing value in the constructor so that
+        // the comparator doesn't check docsWithField since we replace missing values in select()
+        return new IntComparator(numHits, null, null, reversed, false) {
+            @Override
+            public LeafFieldComparator getLeafComparator(LeafReaderContext context) throws IOException {
+                return new IntLeafComparator(context) {
+                    @Override
+                    protected NumericDocValues getNumericDocValues(LeafReaderContext context, String field) throws IOException {
+                        return IntValuesComparatorSource.this.getNumericDocValues(context, iMissingValue);
+                    }
+                };
+            }
+        };
+    }
+
+    @Override
+    public BucketedSort newBucketedSort(
+        BigArrays bigArrays,
+        SortOrder sortOrder,
+        DocValueFormat format,
+        int bucketSize,
+        BucketedSort.ExtraData extra
+    ) {
+        return new BucketedSort.ForInts(bigArrays, sortOrder, format, bucketSize, extra) {
+            private final int iMissingValue = (Integer) missingObject(missingValue, sortOrder == SortOrder.DESC);
+
+            @Override
+            public Leaf forLeaf(LeafReaderContext ctx) throws IOException {
+                return new Leaf(ctx) {
+                    private final NumericDocValues docValues = getNumericDocValues(ctx, iMissingValue);
+                    private long docValue;
```

Comment:
```suggestion
                    private int docValue;
```

Super minor, we could probably store `docValue` as `int` instead of type casting in getter

Reply:
Yeah, one place we need to do at least. I put them in ```advanceExact``` now. This is inline with ```FloatValuesComparatorSource``` now. 
Lucene only provides long/double value reads on NumericDocValues.
This ```FloatValuesComparatorSource``` implementation also had to type caste from double to float. No harm though.

Reply:
It is just less number of conversions (possibly though)